### ISSUE TITLE
Enable Autocenter on digital Joystick

### ISSUE THREAD

**Issue by apison** [libretro/mame2003-plus-libretro#48]
I'm using mame2003-plus in my arcade cabinet but it miss an autocenter feature when using digital arcade joystick that is present in mame2010, mame2000 and fba.
This make unplayble with arcade joystick a lot of game such as outrun, super hang-on, all lightgun game.
Is possible to enable autocenter in mame2003-plus like the others core?

**Comment by arcadez** [comment]
I dunno if it's a case of enabling it as the autocentre code might not be present in this source
but i suppose it would be possible to take it from say mame2000 and port it across to the
mame2003-plus core.

**Comment by apison** [comment]
I'm refering to games that use analog input for example outrun or super hangon , for example when you turn left and then release the stick the car continue to steer left instead to go straight. You have to turn right tryng to center it. This makes the games unplayable, in all the other cores I ve tried when a direction is released the player go straight.

**Comment by grant2258** [comment]
My dragonwise use arcade controller sets to analog by default. I dont have these issues at all.  I would be a nice feature for those that only have digital setups.  Im not really sure whats changed in the input system to be honest but im sure it will be doable

**Comment by apison** [comment]
Thank you grant, for istance in mame2010 analog input menu there is autocenter speed option, if you put is value to 0 you obtain the same behavior of mame2003

**Comment by arcadez** [comment]
If you look at the outrun driver you have the choice of setting analog or digital for the inputs
although both types have this flag for the steering IPF_CENTER which should stop the car
from sticking on the left or right.

/* hang-on's accel/brake are really both analog controls, but I've added them
as digital as well to see what works better */
#define HANGON_DIGITAL_CONTROLS

it's set to digital by default @apison i suppose for testing purposes you could reem out
the above line and then make a build and then try outrun with analog controls to see if
ot works better for ya that way

**Comment by grant2258** [comment]
I think we are going to end up in a whole lot of trouble down between cores line if we start editing driver source files. Look what happened with mame2003 mapped  around sf2

**Comment by apison** [comment]
Sadly im not able to compile under linux and im using it in jamma pandora box 4 board, ill try to learn to compile for this board  if i have time,  maybe i wasnt clear  enought ,im using a digital arcade joystick, are you guys tried these game with digital controller or analog?  Are they work correctly for you.

**Comment by markwkidd** [comment]
@SapphireDrew I am starting to get into a habit of saying it, but is there any way you have a chance to test this?

**Comment by SapphireTactician** [comment]
I don't own an arcade stick, nor do I own the board the OP has mentioned, but if it'll help, I can try with my Linux PC and a regular controller that has analog control sticks like the SNES30 if that will help.

**Comment by SapphireTactician** [comment]
@markwkidd @arcadez So if I understand correctly, I need to remove the following code:

```
/* hang-on's accel/brake are really both analog controls, but I've added them
as digital as well to see what works better */
#define HANGON_DIGITAL_CONTROLS
```

Then recompile and test with an analog controller? Like I said, I don't have an arcade stick, will the 8bitdo SNES30 suffice?

**Comment by arcadez** [comment]
just reem it out to disable it

//#define HANGON_DIGITAL_CONTROLS

**Comment by SapphireTactician** [comment]
@markwkidd @arcadez I tested `outrun` and made the mentioned change to `src/drivers/outrun.c`. I bisected this with the latest release of MAME 2003 Plus via RA's Online Updater. I could not reproduce on my Linux distro with my analog controller. It may not be possible to reproduce the bug without OP's exact hardware setup since he said he's using an actual arcade cabinet.

With both builds however, I did observe that in the `Controls` section of the Quick Menu, that if `Joystick Up`, `Joystick Down`, `Joystick Left`, and `Joystick Right` are assigned to the Left or Right Analog Axises instead of the D-Pad with the Analog to Digital Type setting enabled, that the car would always swerve to the right when neutral. Perhaps that is what @apison is referring to?

Please advise.

**Comment by markwkidd** [comment]
@grant2258 do you know if this is still an issue? I would be glad to try to find someone else as a tester if it still might be.

**Comment by grant2258** [comment]
I fixed out analog controls up quite a bit so we it doesnt lock out using digital controls. So buttons do work for us now as it was locked out before like off road racer ect. I did leave a message in the retropie forums on how to fix this in mame2003.

I think what the op is experiencing is a zero delay controller in analog mode. He would need to put his zero delay control in digital mode .  when the controller is in analog mode it either full on or off. It wont be an issue is the zero delay controller was in digital mode
